lib/better_html/tokenizer/location.rb

```diff
@@ -1,76 +1,59 @@
+require 'parser/source/buffer'
+require 'parser/source/range'
+
 module BetterHtml
   module Tokenizer
-    class Location
-      attr_accessor :document, :start, :stop
-
-      def initialize(document, start, stop)
-        raise ArgumentError, "start location #{start} is out of range for document of size #{document.size}" if start > document.size
-        raise ArgumentError, "stop location #{stop} is out of range for document of size #{document.size}" if stop > document.size
-        raise ArgumentError, "end of range must be greater than start of range (#{stop} < #{start})" if stop < start
+    class Location < ::Parser::Source::Range
+      def initialize(buffer, begin_pos, end_pos)
+        raise ArgumentError, 'first argument must be Parser::Source::Buffer' unless buffer.is_a?(::Parser::Source::Buffer)
+        raise ArgumentError, "begin_pos location #{begin_pos} is out of range for document of size #{buffer.source.size}" if begin_pos > buffer.source.size
+        raise ArgumentError, "end_pos location #{end_pos} is out of range for document of size #{buffer.source.size}" if (end_pos - 1) > buffer.source.size
 
-        @document = document
-        @start = start
-        @stop = stop
+        super(buffer, begin_pos, end_pos)
       end
 
       def range
-        Range.new(start, stop)
+        Range.new(begin_pos, end_pos, true)
       end
 
       def line_range
         Range.new(start_line, stop_line)
       end
 
-      def source
-        @document[range]
-      end
-
-      def start_line
-        @start_line ||= calculate_line(start)
-      end
-
-      def line
-        start_line
-      end
-
-      def stop_line
-        @stop_line ||= calculate_line(stop)
-      end
+      alias_method :start_line, :line
+      alias_method :stop_line, :last_line
+      alias_method :start_column, :column
+      alias_method :stop_column, :last_column
 
-      def start_column
-        @start_column ||= calculate_column(start)
+      def line_source_with_underline
+        spaces = source_line.scan(/\A\s*/).first
+        column_without_spaces = [column - spaces.length, 0].max
+        underscore_length = [[end_pos - begin_pos, source_line.length - column_without_spaces].min, 1].max
+        "#{source_line.gsub(/\A\s*/, '')}\n#{' ' * column_without_spaces}#{'^' * underscore_length}"
       end
 
-      def column
-        start_column
+      def with(begin_pos: @begin_pos, end_pos: @end_pos)
+        self.class.new(@source_buffer, begin_pos, end_pos)
```

Comment:
these helper methods exist in `Parser::Source::Range` but create the wrong class type: https://github.com/whitequark/parser/blob/master/lib/parser/source/range.rb#L178
also that code has not made it to a gem release yet.